src/lifecycle/property.js

```diff
@@ -0,0 +1,164 @@
+    info.updatingProperty = false;
+
+    // TODO Refactor
+    if (info.linkedAttribute) {
+      if (!info.attributeMap) {
+        info.attributeMap = {};
+
+        elem.removeAttribute = function (attrName) {
+          info.removeAttribute.call(this, attrName);
+          if (attrName in info.attributeMap) {
+            elem[info.attributeMap[attrName]] = undefined;
+          }
+        };
+
+        elem.setAttribute = function (attrName, attrValue) {
+          info.setAttribute.call(this, attrName, attrValue);
+          if (attrName in info.attributeMap) {
+            // Could also call getAttribute() but this does the same thing.
+            attrValue = String(attrValue);
+            elem[info.attributeMap[attrName]] = opts.deserialize ? opts.deserialize(attrValue) : attrValue;
+          }
+        };
+      }
+
+      info.attributeMap[info.linkedAttribute] = name;
+    }
+
+    if (initialValue === undefined) {
+      if (info.linkedAttribute && elem.hasAttribute(info.linkedAttribute)) {
+        let attributeValue = elem.getAttribute(info.linkedAttribute);
+        initialValue = opts.deserialize ? opts.deserialize(attributeValue) : attributeValue;
+      } else if (typeof opts.default === 'function') {
+        initialValue = opts.default();
+      } else if (typeof opts.default !== 'undefined') {
+        initialValue = opts.default;
+      }
+    }
+
+    info.internalValue = initialValue;
+  };
+
+  prop.get = function () {
+    let value = opts.get ? opts.get(this) : data(this, `api/property/${name}`).internalValue;
+
+    if (value === undefined && opts.default !== undefined) {
+      if (typeof opts.default === 'function') {
+        value = opts.default();
+      } else {
+        value = opts.default;
+      }
+    }
+
+    return value;
+  };
+
+  prop.ready = function (elem, value) {
+    if (opts.type) {
+      opts.type(value);
+    }
+
+    if (opts.set) {
+      opts.set(elem, {
+        name: name,
+        newValue: value,
+        oldValue: undefined
+      });
+    }
+  };
+
+  prop.set = function (newValue) {
+    let info = data(this, `api/property/${name}`);
+
+    if (info.updatingProperty) {
+      return;
+    }
+
+    info.updatingProperty = true;
+    let oldValue = this[name];
+
+    if (!opts.get) {
+      info.internalValue = newValue;
+    }
+
+    if (newValue === oldValue) {
+      info.updatingProperty = false;
+      return;
+    }
+
+    if (opts.type) {
+      opts.type(newValue);
+    }
+
+    if (info.linkedAttribute) {
+      let serializedValue = opts.serialize ? opts.serialize(newValue) : newValue;
+      if (serializedValue === undefined) {
+        info.removeAttribute.call(this, info.linkedAttribute);
+      } else {
+        info.setAttribute.call(this, info.linkedAttribute, serializedValue);
+      }
+    }
+
+    let changeData = {
+      name: name,
+      newValue: newValue,
+      oldValue: oldValue
+    };
+
+    if (opts.set) {
+      opts.set(this, changeData);
```

Comment:
I think it's fine for the element to be the first argument and to not set `this`.

In my initial question in this thread I overlooked the fact that if you want to pass descriptors rather than opt-in to the functionality of this file you just need to provide a function for the property definition in skate (rather than an object with `default`, `emit`, etc.), so it really isn't a problem that `set` in those two approaches has a different signature.